sled-agent/src/bootstrap/server.rs

```diff
@@ -688,41 +553,27 @@ impl Inner {
         }
     }
 
-    async fn handle_hardware_update(
-        &self,
-        hardware_update: Result<HardwareUpdate, broadcast::error::RecvError>,
-        log: &Logger,
-    ) {
-        info!(
-            log,
-            "Handling hardware update message";
-            "phase" => "bootstore-steady-state",
-            "update" => ?hardware_update,
-        );
-
-        self.managers
-            .handle_hardware_update(
-                hardware_update,
-                self.state.sled_agent(),
-                &log,
-            )
-            .await;
-    }
-
     async fn handle_start_sled_agent_request(
         &mut self,
         request: StartSledAgentRequest,
         response_tx: oneshot::Sender<Result<SledAgentResponse, String>>,
         log: &Logger,
     ) {
-        let request_id = request.body.id;
-        match &self.state {
-            SledAgentState::Bootstrapping => {
+        match &mut self.state {
+            SledAgentState::Bootstrapping(
+                sled_agent_started_tx,
+                underlay_available_tx,
+            ) => {
+                let request_id = request.body.id;
+                // Extract from an option to satisfy the borrow checker
```

Comment:
This is absolutely not safe. @jgallagher and I talked about this and it's also non-trivial to make it such that `start_sled_agent` can fail at all and the system still works. We need to rework the persistent state management to allow that to work, and that is going to be very tricky to get right. This also affects RSS since the code to add a sled reuses the bulk of the RSS path. 

In the short term I'm going to change [this error path to panic explicitly](https://github.com/oxidecomputer/omicron/blob/713243a8bb5390851f399d0d1d28401d9d6489f8/sled-agent/src/bootstrap/server.rs#L594) and add some comments around why, and open an issue for the larger problem of persistent state / idempotence.

Reply:
Documented and issue opened in [cfc3ef7](https://github.com/oxidecomputer/omicron/pull/4332/commits/cfc3ef7c732afdfbf476ac4a62b64471c697bfbb)